frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCache.java

```diff
@@ -106,6 +103,30 @@ public static MetricCache getInstance() {
         return instance;
     }
 
+    public IMetric addAutoDetectMetricBackend(Metric parsedMetric) {
+        List<String> dimensionNames = new ArrayList<String>();
+        for (Dimension dimension : parsedMetric.getDimensions()) {
+            dimensionNames.add(dimension.getName());
+        }
+
```

Comment:
`parsedMetric` is an instance of [Metric](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Metric.java#L27) object but while creating a new metric cache object using `MetricBuilder.build`, we need to provide a list of dimension names, hence the new list.

No, the return value of `getDimensions` is not immutable.

Reply:
Then, why parsedMetric.getDimensions().add("Hostname") can not work?

Reply:
`parsedMetric` is an instance of [Metric](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Metric.java#L27). `parsedMetric.getDimensions()` returns a list of [Dimension](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Dimension.java) objects.

- When [registering a metric to cache](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/MetricBuilder.java#L25), we need to provide a list of dimension names. So, we need to extract the dimension names from the [Dimension](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Dimension.java) objects.
- When [updating a metric](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/LogMetric.java#L34), we need to provide a list of dimension values. So, we need to extract the dimension values from the [Dimension](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Dimension.java) objects.
 
Therefore the new lists are being created above.

Also, the [Metric](https://github.com/pytorch/serve/blob/00609b0474ca376369901e2bd511d7bbbb9074a8/frontend/server/src/main/java/org/pytorch/serve/metrics/Metric.java#L95-L97) class API returns references to private objects which should ideally be immutable. So, I believe it is better to create a new list like we are doing here.

Reply:
it takes time to deepcopy. Metric object will be GC eventually if it is just used once at here and not referenced any more. The shallow copy of dimension name is fine and faster if Metric is only used at here for building IMetrics and it will not be GC.

Reply:
Added `getDimensionNames` and `getDimensionValues` utility functions to [Metric](https://github.com/pytorch/serve/pull/2769/files#diff-fb226a2a07fa69a6476ec5fe0cb7467c3e5b82ba1193668127b88bc46e400963R118-R124) class.